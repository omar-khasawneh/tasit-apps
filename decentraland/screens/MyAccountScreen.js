import React from "react";
import MyAccount from "@presentational/MyAccount";
import { connect } from "react-redux";
import AccountCreationActions from "@constants/AccountCreationActions";
import { MISSING, DONE } from "@constants/ActionStatus";
import PropTypes from "prop-types";

export class MyAccountScreen extends React.Component {
  render() {
    const { accountInfo } = this.props;
    const { creationActions, account } = accountInfo;
    const isAccountCreated = account !== null;

    const creationSteps = [];

    creationSteps.push({
      name: "Account created",
      action: null,
      status: isAccountCreated ? DONE : MISSING,
    });

    Object.keys(AccountCreationActions).forEach(creationStatus => {
      const creationAction = {
        name: AccountCreationActions[creationStatus].name,
        action: creationStatus,
      };
      // TODO: As soon as we store action status in redux, this logic will change
      // transaction pending, confirmed once, confirmed many times, failed, etc.
      if (creationActions.hasOwnProperty(creationStatus)) {
        creationAction.status = DONE;
      } else {
        creationAction.status = MISSING;
      }
      creationSteps.push(creationAction);
    });

    return (
      <MyAccount
        progress={this._getPercentage(isAccountCreated, creationSteps)}
        creationActions={creationSteps}
      />
    );
  }

  _getPercentage(isAccountCreated, creationSteps) {
    let percentage = isAccountCreated ? 0.25 : 0;
    creationSteps.forEach(creationStep => {
      const { status, action } = creationStep;
      const isDone = status === DONE;
      const hasDefinedActionPercentage = AccountCreationActions.hasOwnProperty(
        action
      );
      if (isDone && hasDefinedActionPercentage) {
        percentage += AccountCreationActions[action].percentage;
      }
    });
    return percentage;
  }
}

MyAccountScreen.propTypes = {
  accountInfo: PropTypes.object,
};

const mapStateToProps = state => {
  const { accountInfo } = state;
  return { accountInfo };
};

export default connect(mapStateToProps)(MyAccountScreen);
