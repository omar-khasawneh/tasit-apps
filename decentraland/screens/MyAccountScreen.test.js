import React from "react";
import { shallow } from "enzyme";
import { MyAccountScreen } from "./MyAccountScreen";
import {
  FUNDING_WITH_ETH,
  FUNDING_WITH_MANA,
  APPROVING_MARKETPLACE,
} from "@constants/AccountCreationStatus";

describe("MyAccountScreen", () => {
  describe("renders the component", () => {
    it("initial state (without account)", async () => {
      expect(
        shallow(
          <MyAccountScreen
            accountInfo={{
              account: null,
              creationActions: {},
            }}
          />
        )
      ).toMatchSnapshot();
    });

    it("funding with ethers", async () => {
      expect(
        shallow(
          <MyAccountScreen
            accountInfo={{
              account: {},
              creationActions: {
                [FUNDING_WITH_ETH]: {},
              },
            }}
          />
        )
      ).toMatchSnapshot();
    });

    it("funding with ethers with null action (used when action is unknowed)", async () => {
      expect(
        shallow(
          <MyAccountScreen
            accountInfo={{
              account: {},
              creationActions: {
                [FUNDING_WITH_ETH]: null,
              },
            }}
          />
        )
      ).toMatchSnapshot();
    });

    it("funding with mana AND approving marketplace", async () => {
      expect(
        shallow(
          <MyAccountScreen
            accountInfo={{
              account: {},
              creationActions: {
                [FUNDING_WITH_ETH]: {},
                [FUNDING_WITH_MANA]: {},
                [APPROVING_MARKETPLACE]: {},
              },
            }}
          />
        )
      ).toMatchSnapshot();
    });
  });
});
