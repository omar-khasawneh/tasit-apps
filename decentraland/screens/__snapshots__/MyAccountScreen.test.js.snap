// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyAccountScreen renders the component with account 1`] = `
<MyAccount
  creationSteps={
    Array [
      Object {
        "creationStatus": null,
        "name": "Account created",
        "percentage": 0.25,
        "status": "done",
      },
      Object {
        "creationStatus": "FUNDING_WITH_ETH",
        "name": "Funded with ETH",
        "percentage": 0.25,
        "status": "done",
      },
      Object {
        "creationStatus": "FUNDING_WITH_MANA",
        "name": "Funded with MANA tokens",
        "percentage": 0.25,
        "status": "missing",
      },
      Object {
        "creationStatus": "APPROVING_MARKETPLACE",
        "name": "Linked to marketplace",
        "percentage": 0.25,
        "status": "missing",
      },
    ]
  }
  progress={0.5}
/>
`;

exports[`MyAccountScreen renders the component without account 1`] = `
<MyAccount
  creationSteps={
    Array [
      Object {
        "creationStatus": null,
        "name": "Account created",
        "percentage": 0.25,
        "status": "missing",
      },
      Object {
        "creationStatus": "FUNDING_WITH_ETH",
        "name": "Funded with ETH",
        "percentage": 0.25,
        "status": "missing",
      },
      Object {
        "creationStatus": "FUNDING_WITH_MANA",
        "name": "Funded with MANA tokens",
        "percentage": 0.25,
        "status": "missing",
      },
      Object {
        "creationStatus": "APPROVING_MARKETPLACE",
        "name": "Linked to marketplace",
        "percentage": 0.25,
        "status": "missing",
      },
    ]
  }
  progress={0}
/>
`;
